src/cloudWatchLogs/changeLogSearch.ts

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as telemetry from '../shared/telemetry/telemetry'
+import * as vscode from 'vscode'
+import { showInputBox } from '../shared/ui/inputPrompter'
+import { isLogStream } from './cloudWatchLogsUtils'
+import { prepareDocument } from './commands/searchLogGroup'
+import { getActiveUri } from './document/logStreamDocumentProvider'
+import { CloudWatchLogsData, filterLogEventsFromUriComponents, LogStreamRegistry } from './registry/logStreamRegistry'
+import { isViewAllEvents, TimeFilterResponse, TimeFilterSubmenu } from './timeFilterSubmenu'
+
+type ChangeableParam = 'filterPattern' | 'timeFilter'
+
+export async function getNewData(
+    registry: LogStreamRegistry,
+    param: ChangeableParam,
+    oldUri: vscode.Uri
+): Promise<CloudWatchLogsData | undefined> {
+    const oldData = registry.getLogData(oldUri) as CloudWatchLogsData
+    const newData: CloudWatchLogsData = {
+        ...oldData,
+        data: [],
+        next: undefined,
+        previous: undefined,
+    }
+    let newPattern: string | undefined
+    let newTimeRange: TimeFilterResponse | undefined
+
+    switch (param) {
+        case 'filterPattern':
+            newPattern = await showInputBox({
+                title: isLogStream(oldUri) ? 'Filter Log Stream Results' : 'Log Group Keyword Search',
+                placeholder: oldData.parameters.filterPattern ?? 'Enter Text Here',
+            })
+            if (newPattern === undefined) {
+                return
```

Comment:
Are there any general 'rules' of when to provide a docstring or is it more a judgement call on whether or not the function name provides all the context necessary?

Reply:
Can't say there's really any rules, it's mostly a judgement call. You can add docstrings whenever you want but they should say something that is not obvious just from looking at the code. Sometimes it's better to rename the function instead. For example, if we called the function `promptForNewData` then it's a lot more obvious that this is asking the user for something, not just 'getting' the data from wherever. Maybe a good 'rule' is that if the function causes "side-effects" (e.g. changes the UI) then more descriptive names are preferred.

Keep in mind that the more 'visible' something is, the more important the name. Things like private methods/variables/etc. don't matter as much because they won't be seen as much. Terse names are a good thing as long as they won't cause confusion.

Reply:
+1, but also, if I find myself thinking about quirks or special knowledge, I add that to the docstring.